### DISCUSSION TITLE
Import errors due to `client/registerCapability` failed for a go project

### DISCUSSION THREAD

**Discussion by zhengyangfeng00** [joaotavora/eglot#1267] [Q&A]
When I opened a file in the project and started eglot, I got bunch of import errors. This project has go mod enabled. Rougly the directory structure is:

```
root/
  .git/
   goproject_root/
     go.mod
      ...
```

I tried to reproduce this issue on a minimum project that has the same directory structure, but it worked on this simplified project.

The (simplified) project layout is:

```
eglottest
|-- .git
└── goproject
    ├── go.mod
    ├── go.sum
    ├── main.go
    └── vendor
        ├── github.com
        │   └── davecgh
        │       └── go-spew
        │           ├── LICENSE
        │           └── spew
        │               ├── bypass.go
        │               ├── bypasssafe.go
        │               ├── common.go
        │               ├── config.go
        │               ├── doc.go
        │               ├── dump.go
        │               ├── format.go
        │               └── spew.go
        └── modules.txt
```

By comparing the difference in the `*EGLOT*` buffer's output, I was able to pin the issue to an error.

Here is what didn't work(note it doesn't add vendor directory):

```
(:jsonrpc "2.0" :method "client/registerCapability" :params
          (:registrations
           [(:id "workspace/didChangeWatchedFiles-0" :method "workspace/didChangeWatchedFiles" :registerOptions
                 (:watchers
                  [(:globPattern "**/*.{go,mod,sum,work}" :kind 7)]))])
          :id 3)
[client-reply] (id:3) ERROR Tue Aug  8 15:44:49 2023:
(:jsonrpc "2.0" :id 3 :error
          (:code -32603 :message "Internal error"))
...
```

Here is what worked:

```
(:jsonrpc "2.0" :method "client/registerCapability" :params
          (:registrations
           [(:id "workspace/didChangeWatchedFiles-0" :method "workspace/didChangeWatchedFiles" :registerOptions
                 (:watchers
                  [(:globPattern "**/*.{go,mod,sum,work}" :kind 7)
                   (:globPattern "{/Users/zhengyangfeng/eglottest/goproject/vendor,/Users/zhengyangfeng/eglottest/goproject/vendor/github.com,/Users/zhengyangfeng/eglottest/goproject/vendor/github.com/davecgh,/Users/zhengyangfeng/eglottest/goproject/vendor/github.com/davecgh/go-spew,/Users/zhengyangfeng/eglottest/goproject/vendor/github.com/davecgh/go-spew/spew}" :kind 7)]))])
          :id 3)
[client-reply] (id:3) Tue Aug  8 15:44:05 2023:
...
```

In the log with the error, I expect eglot to add vendor directory like the project that worked. Could anyone give advice on how to debug further?

**Comment by joaotavora** [comment]
Sounds similar to https://github.com/joaotavora/eglot/issues/1258.

Is your "real project" very large? How many `**/*.{go,mod,sum,work}` files do you have under it? Are you using Mac OSX?

**Comment by zhengyangfeng00** [comment]
I've solved my issue. Essentially I just need to mark the subdirectory I wanted to work on as a project.

In my case, I need to create a `.projectile` file in that sub-directory. As @joaotavora said eglot mode uses `project.el` to find the project root. However I have also installed `projectile.el`, which integrates itself with `project.el`. This can be verified by looking at the variable doc for `project-find-functions`, note there is a `project-projectile`:

> project-find-functions is a variable defined in ‘project.el’.
>
> Its value is (project-projectile project-try-vc)

Projectile allows users to manually mark a directory as a project by creating an empty `.projectile` in that directory.

Once that is done, I confirmed the project root was set correctly by visiting one file and run `(project-current)`. Then I started `eglot` and everything works!

Thanks @joaotavora for pointing to the right direction. I really like the way things work now as it is a scalable setup for working in a monorepo.

For completeness, here are some additional information for my setup:

1. I'm using MacOSX and use nix to manage emacs (don't think this matters)
2. re: how many files in the project

    Here is the go.mod files from the root:
    
    ```
    ❯ find . -type f -name "go.mod"
    ./light-clients/altair/go.mod
    ./light-clients/parlia/go.mod
    ./light-clients/parlia/plugin/go.mod
    ./light-clients/bor/go.mod
    ./light-clients/ethclique/go.mod
    ./light-clients/bnbtm/go.mod
    ./chain/go.mod
    ```
    
    The `chain/` directory contains the actual code that I want to work on, and it has a vendor directory.
